### DISCUSSION TITLE
Best Practices for RemoteDatabase in Kotlin (specifically thread-safety)

### DISCUSSION THREAD

**Discussion by tristanradams** [ArcadeData/arcadedb#1599] [Q&A]
I am currently in the process of trying to convert my Spring Kotlin project to using ArcadeDB for its graph database. In the process of setting things up, I noticed that the `RemoteDatabase` is not thread safe and that I should produce one per thread. 

Here is my rough draft so far of a possible wrapper class (which my application would only have one instance of the wrapper class) I could use to keep things safe:
```
import com.arcadedb.remote.RemoteDatabase

class RemoteDatabaseThreadProxy(
    private val host: String,
    private val port: Int,
    private val databaseName: String,
    private val username: String,
    private val password: String
) {
    private val remoteDatabaseByThreadName = mutableMapOf<String, RemoteDatabase>()

    fun get(): RemoteDatabase {
        val currentThread = Thread.currentThread()
        return remoteDatabaseByThreadName.getOrPut(currentThread.name) {
            RemoteDatabase(host, port, databaseName, username, password)
        }
    }
}
```

My main questions are:
- Would this wrapper class be appropriate for getting an instance of `RemoteDatabase`? 
- Is it an "expensive" transaction to fully initialize a `RemoteDatabase` if many threads are being created?
- Would the approach above cause any memory leaks since I am not expressly closing the `RemoteDatabase` instances when `Thread`s are closed? 

Please let me know if any more context is needed or if I am shooting in the completely wrong direction!

**Comment by jstuyts** [comment]
You should take a look at [`ThreadLocal`](https://docs.oracle.com/javase/8/docs/api/java/lang/ThreadLocal.html). Using [`withInitial(...)`](https://docs.oracle.com/javase/8/docs/api/java/lang/ThreadLocal.html#withInitial-java.util.function.Supplier-) you can ensure that a remote database is available the first time it is used.
